core/src/main/java/com/graphhopper/reader/dem/AbstractEdgeElevationInterpolator.java

```diff
@@ -0,0 +1,114 @@
+
+	private final GraphHopperStorage storage;
+	protected final DataFlagEncoder dataFlagEncoder;
+	private final NodeElevationInterpolator nodeElevationInterpolator;
+	private final ElevationInterpolator elevationInterpolator = new ElevationInterpolator();
+
+	public AbstractEdgeElevationInterpolator(GraphHopperStorage storage, DataFlagEncoder dataFlagEncoder) {
+		this.storage = storage;
+		this.dataFlagEncoder = dataFlagEncoder;
+		this.nodeElevationInterpolator = new NodeElevationInterpolator(storage);
+	}
+
+	protected abstract boolean isStructureEdge(EdgeIteratorState edge);
+	
+	public GraphHopperStorage getStorage() {
+		return storage;
+	}
+
+	public void execute() {
+		interpolateElevationsOfTowerNodes();
+		interpolateElevationsOfPillarNodes();
+	}
+
+	private void interpolateElevationsOfTowerNodes() {
+		final AllEdgesIterator edge = storage.getAllEdges();
+		final GHBitSet visitedEdgeIds = new GHBitSetImpl(edge.getMaxId());
+		final EdgeExplorer edgeExplorer = storage.createEdgeExplorer();
+
+		while (edge.next()) {
+			final int edgeId = edge.getEdge();
+			if (isStructureEdge(edge)) {
+				if (!visitedEdgeIds.contains(edgeId)) {
+					processStructureEdge(edge, visitedEdgeIds, edgeExplorer, edgeId);
+				}
+			}
+			visitedEdgeIds.add(edgeId);
+		}
+	}
+
+	private void processStructureEdge(final EdgeIteratorState structureEdge, final GHBitSet visitedEdgeIds,
+			final EdgeExplorer edgeExplorer, int edgeId) {
+		final TIntSet outerNodeIds = new TIntHashSet();
+		final TIntSet innerNodeIds = new TIntHashSet();
+		gatherOuterAndInnerNodeIdsOfStructure(edgeExplorer, structureEdge, visitedEdgeIds, outerNodeIds,
+				innerNodeIds);
+		nodeElevationInterpolator.interpolateElevationsOfInnerNodes(outerNodeIds.toArray(), innerNodeIds.toArray());
+	}
+
+	public void gatherOuterAndInnerNodeIdsOfStructure(final EdgeExplorer edgeExplorer,
+			final EdgeIteratorState structureEdge, final GHBitSet visitedEdgesIds, final TIntSet outerNodeIds,
+			final TIntSet innerNodeIds) {
+		final BreadthFirstSearch gatherOuterAndInnerNodeIdsOfStructureSearch = new BreadthFirstSearch() {
+			protected boolean checkAdjacent(EdgeIteratorState edge) {
+				visitedEdgesIds.add(edge.getEdge());
+				final int baseNodeId = edge.getBaseNode();
+				boolean isStructureEdge = isStructureEdge(edge);
+				if (!isStructureEdge) {
+					innerNodeIds.remove(baseNodeId);
+					outerNodeIds.add(baseNodeId);
+				} else if (!outerNodeIds.contains(baseNodeId)) {
+					innerNodeIds.add(baseNodeId);
+				}
+				return isStructureEdge;
+			}
+		};
+		gatherOuterAndInnerNodeIdsOfStructureSearch.start(edgeExplorer, structureEdge.getBaseNode());
+	}
+
+	private void interpolateElevationsOfPillarNodes() {
+		final EdgeIterator edge = storage.getAllEdges();
+		while (edge.next()) {
+			if (isStructureEdge(edge)) {
+				int firstNodeId = edge.getBaseNode();
+				int secondNodeId = edge.getAdjNode();
+
+				double lat0 = storage.getNodeAccess().getLat(firstNodeId);
```

Comment:
I would assign nodeAccess to a local variable outside of the loop. Potentially the JVM is already doing this but it also reads a bit faster :)